endpoints/setuid.go

```diff
@@ -175,19 +189,151 @@ func NewSetUIDEndpoint(cfg *config.Configuration, syncersByBidder map[string]use
+		reqInfo.GDPRSignal = legacySignal
+		reqInfo.Consent = legacyConsent
+	}
+
+	if reqInfo.Consent == "" && reqInfo.GDPRSignal == gdpr.SignalYes {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, errors.New("GDPR consent is required when gdpr signal equals 1")
+	}
+
+	return reqInfo, err
+}
+
+// parseGDPRFromGPP parses and validates the "gpp_sid" and "gpp" query fields.
+func parseGDPRFromGPP(query url.Values) (gdpr.RequestInfo, error) {
+	var gdprSignal gdpr.Signal = gdpr.SignalAmbiguous
+	var gdprConsent string = ""
+	var err error
+
+	gdprSignal, err = parseSignalFromGppSidStr(query.Get("gpp_sid"))
+	if err != nil {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	gdprConsent, err = parseConsentFromGppStr(query.Get("gpp"))
+	if err != nil {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	return gdpr.RequestInfo{
+		Consent:    gdprConsent,
+		GDPRSignal: gdprSignal,
+	}, nil
+}
+
+// parseLegacyGDPRFields parses and validates the "gdpr" and "gdpr_consent" query fields which
+// are considered deprecated in favor of the "gpp" and "gpp_sid". The parsed and validated GDPR
+// values contained in "gpp" and "gpp_sid" are passed in the parameters gppGDPRSignal and
+// gppGDPRConsent. If the GPP parameters come with non-default values, this function discards
+// "gdpr" and "gdpr_consent" and returns a warning.
+func parseLegacyGDPRFields(query url.Values, gppGDPRSignal gdpr.Signal, gppGDPRConsent string) (gdpr.Signal, string, error) {
+	var gdprSignal gdpr.Signal = gdpr.SignalAmbiguous
+	var gdprConsent string
+	var warning error
+
+	if gdprQuerySignal := query.Get("gdpr"); len(gdprQuerySignal) > 0 {
+		if gppGDPRSignal == gdpr.SignalAmbiguous {
+			switch gdprQuerySignal {
+			case "0":
+				fallthrough
+			case "1":
+				if zeroOrOne, err := strconv.Atoi(gdprQuerySignal); err == nil {
+					gdprSignal = gdpr.Signal(zeroOrOne)
+				}
+			default:
+				return gdpr.SignalAmbiguous, "", errors.New("the gdpr query param must be either 0 or 1. You gave " + gdprQuerySignal)
+			}
+		} else {
+			warning = &errortypes.Warning{
+				Message:     "'gpp_sid' signal value will be used over the one found in the deprecated 'gdpr' field.",
+				WarningCode: errortypes.UnknownWarningCode,
+			}
+		}
+	}
+
+	if gdprLegacyConsent := query.Get("gdpr_consent"); len(gdprLegacyConsent) > 0 {
+		if len(gppGDPRConsent) > 0 {
+			warning = &errortypes.Warning{
+				Message:     "'gpp' value will be used over the one found in the deprecated 'gdpr_consent' field.",
+				WarningCode: errortypes.UnknownWarningCode,
+			}
+		} else {
+			gdprConsent = gdprLegacyConsent
+		}
+	}
+	return gdprSignal, gdprConsent, warning
+}
+
+func parseSignalFromGppSidStr(strSID string) (gdpr.Signal, error) {
+	gdprSignal := gdpr.SignalAmbiguous
+
+	if len(strSID) > 0 {
+		gppSID, err := stringutil.StrToInt8Slice(strSID)
+		if err != nil {
+			return gdpr.SignalAmbiguous, fmt.Errorf("Error parsing gpp_sid %s", err.Error())
```

Comment:
For my own understanding is it okay to use fmt when returning an error? I typically use something like `errors.New()`

Reply:
I use fmt.Error when I want to pass a parameter to error message. For instance when you want to build dynamic error messages and use `%s`, `%d` to format values. fmt.Error returns usual error, it just helps to format the message.